pcsx2/DEV9/AdapterUtils.cpp

```diff
@@ -244,30 +247,69 @@ std::optional<MAC_Address> AdapterUtils::GetAdapterMAC(Adapter* adapter)
 
 	return std::nullopt;
 }
-#elif defined(__POSIX__)
-#ifdef __linux__
+#else
 std::optional<MAC_Address> AdapterUtils::GetAdapterMAC(Adapter* adapter)
 {
-	struct ifreq ifr;
-	strcpy(ifr.ifr_name, adapter->ifa_name);
+	MAC_Address macAddr = {};
+#if defined(AF_LINK)
+	ifaddrs* adapterInfo;
+	const int error = getifaddrs(&adapterInfo);
+	if (error)
+	{
+		Console.Error("DEV9: Failed to get adapter information %s", error);
+		return std::nullopt;
+	}
 
-	int fd = socket(AF_INET, SOCK_DGRAM, 0);
-	int ret = ioctl(fd, SIOCGIFHWADDR, &ifr);
-	close(fd);
+	const ScopedGuard adapterInfoGuard = [&adapterInfo]() {
+		freeifaddrs(adapterInfo);
+	};
 
-	if (ret == 0)
-		return *(MAC_Address*)ifr.ifr_hwaddr.sa_data;
+	for (const ifaddrs* po = adapterInfo; po; po = po->ifa_next)
+	{
+		if (strcmp(po->ifa_name, adapter->ifa_name))
+			continue;
 
-	return std::nullopt;
-}
+		if (po->ifa_addr->sa_family != AF_LINK)
+			continue;
+
+		// We have a valid MAC address.
+		std::memcpy(&macAddr, LLADDR(reinterpret_cast<sockaddr_dl*>(po->ifa_addr)), sizeof(macAddr));
+		return macAddr;
+	}
+	Console.Error("DEV9: Failed to get MAC address for adapter using AF_LINK");
 #else
-std::optional<MAC_Address> AdapterUtils::GetAdapterMAC(Adapter* adapter)
-{
-	Console.Error("DEV9: Unsupported OS, can't get MAC address");
+	const int sd = socket(AF_INET, SOCK_DGRAM, IPPROTO_IP);
+	if (sd < 0)
+	{
+		Console.Error("DEV9: Failed to open socket");
+		return std::nullopt;
+	}
+	const ScopedGuard sdGuard = [sd]() {
+		close(sd);
+	};
+	struct ifreq ifr = {};
+	StringUtil::Strlcpy(ifr.ifr_name, adapter->ifa_name, std::size(ifr.ifr_name));
+#if defined(SIOCGIFHWADDR)
+	if (ioctl(sd, SIOCGIFHWADDR, &ifr) < 0)
+	{
+		Console.Error("DEV9: Failed to get MAC address for adapter using SIOCGIFHWADDR");
+		return std::nullopt;
+	}
+	std::memcpy(&macAddr, &ifr.ifr_hwaddr.sa_data, sizeof(macAddr));
+#elif defined(SIOCGENADDR)
+	if (ioctl(sd, SIOCGENADDR, &ifr) < 0)
+	{
+		Console.Error("DEV9: Failed to get MAC address for adapter using SIOCGENADDR");
+		return std::nullopt;
+	}
+	std::memcpy(&macAddr, &ifr.ifr_enaddr, sizeof(macAddr));
+#endif
+	return macAddr;
+#endif
+	Console.Error("DEV9: Unsupported OS, can't get network features");
 	return std::nullopt;
 }
```

Comment:
With how you have the defines setup, the error message will not be hit even if `AF_LINK`, `SIOCGIFHWADDR` & `SIOCGENADDR` where all undefined

Instead you would return a empty mac address

Reply:
Good catch! I will return `macAddr` for both `SIOCGIFHWADDR` and  `SIOCGENADDR`.

Reply:
Pushed! Please take a look

Reply:
Looks good